File: medium/group_anagrams/Go/group_anagrams.go
package main

// Cathegory: Medium

// Success
// Details
// Runtime: 38 ms, faster than 66.98% of Go online submissions (1,723,930) for Group Anagrams.
// Memory Usage: 9.6 MB, less than 8.02% of Go online submissions (1,723,930) for Group Anagrams.

// 118 / 118 test cases passed.
// 	Status: Accepted
// Runtime: 38 ms
// Memory Usage: 9.6 MB

// Copyright 2022 Bogdan Peta, All Rights Reserved

import (
	// "fmt"
	"sort"
)

func groupAnagrams(strs []string) [][]string {

	acc := map[string][]string{}

	for _, w := range strs {
		str_w := []rune(w)
		
		sort.Slice(str_w, func(i int, j int) bool { return str_w[i] < str_w[j] })
		
		w1 := string(str_w)

		acc[w1] = append(acc[w1], w)
	}

	rez := [][]string{}
	for _, v := range acc {
		if v != nil {
			rez = append(rez, v)
		}
	}

	return rez
}

File: medium/group_anagrams/Rust/src/group_anagrams.rs
// Success
// Details
// Runtime: 31 ms, faster than 14.34% of Rust online submissions for Group Anagrams.
// Memory Usage: 6 MB, less than 15.97% of Rust online submissions for Group Anagrams.

// 118 / 118 test cases passed.
// 	Status: Accepted
// Runtime: 31 ms
// Memory Usage: 6 MB

// Date: 9.11.2022

// Copyright 2022 Bogdan Peta, All Rights Reserved

use std::collections::HashMap;

pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>>
{
    let mut acc: HashMap<String, Vec<String>> = HashMap::new();

    for mut w in strs {

        w = w.to_lowercase();
        let mut str_w: Vec<char> = w.chars().collect::<Vec<char>>();
        str_w.sort_by(|s1, s2| s1.cmp(s2));

        let w1 = str_w.into_iter().collect::<String>();
        
        let mut a: Vec<String> = acc.get(&w1).unwrap_or(&vec![]).to_vec();

        a.push(w);
        acc.insert(w1, a);        
    }

    // sort the map to always pass the tests
    let mut sa: Vec::<String> = acc.keys().cloned().collect();
    // sa.sort_by_key(|a| a.to_lowercase());
    sa.sort();

    let mut rez: Vec<Vec<String>> = vec![];

    for val in sa {
        rez.push(acc.get(&val).unwrap().to_vec());
    }
    
    return rez;
}
